File: Pepcoding Classes/Foundation/String/l002string.cpp
#include<iostream>
#include<vector>
using namespace std;

//  Check palindrome
bool isPalindrome(string str){
    int i = 0 ;
    int j = str.size() - 1;
    while(i < j){
        if(str[i++] != str[j--]){
            return false;
        }
    }
    return true;
}

// Print all substrings 
void printAllSubstrings(string str){
    for(int i = 0 ; i < str.size(); i++){
        for(int j = i; j < str.size() ; j++){
            string s = str.substr( i , j + 1 - i );
            cout << s << endl;
        }
    }
}
// -------------------------------------------
void anotherWayToPrintAllSubstrings(string str){
    for(int i = 0 ; i < str.size(); i++){
        for(int len = 1; len + i <= str.size() ; len++){
            string s = str.substr( i , len);
            cout << s << endl;
        }
    }
}

// Print all the Palindromic Substring
void solution(string str) {
    for (int i = 0; i < str.length(); i++) {
        for (int j = i; j < str.length(); j++) {
            string s = str.substr(i, j + 1 - i );
            if (isPalindrome(s)) {
                cout << s << endl;
            }
        }
    }
}

// String compression 
// aaabbccdefff -> a3b2c2def3 
void compression1(string str){
    if(str.length() <= 1){ 
        cout << str ;
        return ;
    }
    string ans = "";
    char prevchar = str[0];
    int i = 1;
    while(i <= str.length()){
        int count = 1;
        while (i < str.length() && prevchar == str[i]){
            count++;
            i++;
        }
        ans += prevchar;
        if(count != 1)  ans += to_string(count); 
        prevchar = str[i];
        i++;
        
    }
    cout << ans << endl; 
}

// aaaaabbbbbcccccdddsefs -> abcdsefs
void compression2(string str){
    string ans = "";
    char prevchar = str[0];
    int i = 1;
    while(i <= str.length()){
        int count = 1;
        while (i < str.length() && prevchar == str[i]){
            count++;
            i++;
        }
        ans += prevchar;
        // if(count != 1)  ans += to_string(count); 
        prevchar = str[i];
        i++;
        
    }
    cout << ans << endl; 
}


// WithoutX2 problem 
string withoutX2(string str){
    string ans = "";
    for(int i = 0 ; i < str.length() ; i++){

        if(i < 2 && str[i] != 'x')   ans += str[i];
        if(i >= 2)  ans += str[i];
    }
    return ans;
}

// permutation

void appendCharInString(string str, char ch,vector<string>& ans){
        
    for(int i=0; i<=str.length();i++){
        string s = str.substr(0,i) + ch + str.substr(i);
        ans.push_back(s);
    }
}
void solution2(string str) {
    // write your code here
    vector<string> ans;
    ans.push_back("");
    for(int i = 0; i< str.length() ; i++){
        char ch = str[i];
        
        vector<string> smallAns;
        for(string s : ans){
            appendCharInString(s,ch,smallAns);
        }
        ans = smallAns;
    }
    for(string s: ans){
        cout << s << " ";
    }
}

// LEETCODE -> Reverse String 
void reverseRange(string& s,int i, int j){
    while(i < j){
        swap(s[i++],s[j--]);
    }
}
string reverseStr(string s, int k) {
    if(k == 0 || k == 1 || s.length() <= 1) return s ;
    
    int i = 0 , n = s.length();
    
    while(i < n){
        if(i + k - 1 < n){
            reverseRange(s,i,i+k-1);
            i += 2*k ;
            
        } else {
            reverseRange(s,i,n-1);
            break;
        }
    }
    return s;
}
// ----------------------------------------------

// First Unique Character in a String 
int firstUniqChar(string s) {
    vector<int> freq(26);
    for(int i=0;i < s.size(); i++){
        freq[s[i] - 'a']++;
    }
    for(int i = 0 ; i < s.size() ; i++){
        if( freq[s[i] - 'a'] == 1){
            return i;
        }
    }
    return -1;
}
// -----------------------------------------------------



// Longest Palindrome
void longestPalindrome(string str) {
    int maxEle = -1e8;
    string ans = ""; 
    for (int i = 0; i < str.length(); i++) {
        for (int j = i; j < str.length(); j++) {
            string s = str.substr(i, j + 1 - i );
            int len = s.length();
            if (isPalindrome(s) && (len > maxEle)) {
                maxEle = len;
                ans = s;
            }
        }
    }
    cout << ans << endl;
}

// ------------------------------------------

// Prime Factorisation and Display the factor with there powers 
#include<iostream>
#include<vector>
using namespace std;

bool isPrime(int n){
    for(int i = 2;i * i <= n;i++){
        if(n % i == 0) return false;
    }
    return true;
}

void primeNumbers(int n,vector<int>& ans){
    for(int i = 2; i * i <= n; i++){
        if(isPrime(i)) ans.push_back(i);
    }
}

void primeFactors(int num,vector<int>& list){

    int idx = 0;
    while(num != 1 && idx < list.size()){
        int count = 0;
        while(num % list[idx] == 0){
            num /= list[idx];
            count++;
        }
        if(count > 0)
           cout << list[idx] << "^" << count << " " ;
        idx++;
    }
    
    if(num > 1)
        cout << num << "^" << 1;
    
    cout << endl;
}

void primeFactorsForQuery(vector<int>& query){
    vector<int> list;
    primeNumbers(10000,list);

    for(int ele : query){
        primeFactors(ele,list);
    }
}

// int main(int argc, char** agrc){
//     //write your code here
//     int n ; cin >> n;
//     vector<int> arr(n);
//     for(int i=0; i< arr.size();i++) cin >> arr[i];
//     primeFactorsForQuery(arr);
// }

// --------------------------------------------------------------------------------------



int main(){
    // compression2("aaaaabbbbbcccccdddsefs");
    // solution2("abc");
    // cout << firstUniqChar("leetcode");

    // string str;
	// cin >> str;
	// longestPalindrome(str);

    // int n ; cin >> n;
    // vector<int> arr(n);
    // for(int i=0; i< arr.size();i++) cin >> arr[i];
    // primeFactorsForQuery(arr);
    cout << reverseStr("opopopopopopopopopop",8);
    return 0;
}

File: Pepcoding Classes/Foundation/String/l002string.java
import java.util.*;
public class l002string{
    public static Scanner scn = new Scanner(System.in);
    public static String consecutiveCharacter(String str){
        if(str.length() <= 1) return str;
        StringBuilder sb = new StringBuilder();

        sb.append(str.charAt(0));
        for(int i = 1; i < str.length(); i++){
            char ch = str.charAt(i);
            char ch0 = str.charAt(i-1);

            sb.append(ch - ch0);
            sb.append(ch);
        }
        return sb.toString();
    }


    public static boolean isPalindrome(String str){
        int i = 0 , j = str.length() -1 ;
        while(i < j){
            if(str.charAt(i++) != str.charAt(j--)){
                return false;
            }
        }
        return true;
    }
    public static void palindromeSubstring(String str){
        for(int i = 0 ; i < str.length(); i++){
            for(int j = i ; j < str.length(); j++){
                String s = str.substring(i,j+1);

                if(isPalindrome(s)){
                    System.out.println(s);
                }
            }    
        }
    }

    public static String toggleString(String str){
        StringBuilder sb = new StringBuilder();
        for(int i = 0 ; i < str.length(); i++){
            char ch = str.charAt(i);

            if(ch >= 'a' && ch <= 'z' ){
                sb.append((char)(ch - 'a' + 'A'));
            }
            else {
                sb.append((char)(ch - 'A' + 'a'));
            }
        }
        return sb.toString();
    }
    public static void anotherWayToPrintAllSubstrings(String str){
        for(int i = 0 ; i < str.length(); i++){
            for(int len = 1; len + i <= str.length() ; len++){
                System.out.println(str.substring( i ,i+ len));
            }
        }
    }

    // String with difference of Everytwo consecutive characters
    public static String consecutivestringdiff(String str){
		int len = str.length();
		StringBuilder ans = new StringBuilder();
		ans.append(str.charAt(0));
		
		for(int i = 1; i < len; i++){
		    char a = str.charAt(i - 1);
		    char b = str.charAt(i);
		    
		    int diff = (int)(b - a);
		    ans.append(diff);
		    ans.append(b);
		}
		
		return ans.toString() ;
	}

    // WithoutX2 problem 
    public static String withoutX2(String str){
        StringBuilder sb = new StringBuilder();
        for(int i = 0 ; i < str.length() ; i++){

            if(i < 2 && str.charAt(i) != 'x')   sb.append(str.charAt(i));
            if(i >= 2)  sb.append(str.charAt(i));
        }
        return sb.toString();
    }
    
   
    // -------------------------------------
    // Add a character at differenet places in string
    public static ArrayList<String> appendCharInString(String str,char ch){
        ArrayList<String> sb = new ArrayList<>();
        for(int i=0; i<=str.length();i++){
            String s = str.substring(0,i) + ch + str.substring(i);
            sb.add(s);
        }
        return sb;
    }
    // ------------------------------------
    
    // permutation
    public static void appendCharInString(String str, char ch,ArrayList<String> ans){
        
        for(int i=0; i<=str.length();i++){
            String s = str.substring(0,i) + ch + str.substring(i);
            ans.add(s);
        }
    }
    public static void solution(String str) {
        // write your code here
        ArrayList<String> ans = new ArrayList<>();
        ans.add("");
        for(int i = 0; i< str.length() ; i++){
            char ch = str.charAt(i);
            
            ArrayList<String> smallAns = new ArrayList<>();
            for(String s : ans){
                appendCharInString(s,ch,smallAns);
            }
            ans = smallAns;
        }
        System.out.println(ans);
    }

    // -----------------------------------

    public static boolean isPrime(int n){
        for(int i = 2;i * i <= n;i++){
            if(n % i == 0) return false;
        }
        return true;
    }

    public static void primeNumbers(int n,ArrayList<Integer> ans){
        for(int i = 2; i * i <= n; i++){
            if(isPrime(i)) ans.add(i);
        }
    }

    public static void primeFactors(int num,ArrayList<Integer> list){

        int idx = 0;
        while(num != 1 && idx < list.size()){
            int count = 0;
            while(num % list.get(idx) == 0){
                num /= list.get(idx);
                count++;
            }
            if(count > 0)
               System.out.print(list.get(idx) + "^" + count + " ");
            idx++;
        }
        
        if(num > 1)
            System.out.print(num + "^" + 1);
        
        System.out.println();
    }

    public static void primeFactorsForQuery(int[] query){
        ArrayList<Integer> list = new ArrayList<>();
        primeNumbers(10000,list);

        for(int ele : query){
            if(ele == 0){
                System.out.println(ele);
                continue;
            } 
            primeFactors(ele,list);
        }
    }
    
    public static void main(String[] args){
        // Scanner scn = new Scanner(System.in);
        // String str = scn.next();
        // solution(str);
        int n = scn.nextInt();  
        int[] arr = new int[n];
        for(int i = 0 ; i < arr.length ; i++) arr[i] = scn.nextInt();
        primeFactorsForQuery(arr);
    }
}